---
title: "76 lines SNP Association to Gut Microbiome Traits During Digestions"
output: md_document
date: "2025-0630"
name: Bo Peng
---


# rMVP

## Input genotype in VCF format
```{r}
library(rMVP)
# Full-featured function (Recommended)
MVP.Data(fileVCF="myVCF.vcf",
         #filePhe="Phenotype.txt",
         fileKin=FALSE,
         filePC=FALSE,
         out="mvp.vcf"
         )

# Only convert genotypes
MVP.Data.VCF2MVP("myVCF.vcf", out='mvp') # the genotype data should be fully imputed before using this function

# calculate from mvp_geno_file
MVP.Data.PC(TRUE, mvp_prefix='mvp.plink', pcs.keep=5)
# calculate from mvp_geno_file
MVP.Data.Kin(TRUE, mvp_prefix='mvp.plink', out='mvp.plink')

Kinship <- attach.big.matrix("mvp.plink.kin.desc")
genotype <- attach.big.matrix("mvp.plink.geno.desc")
phenotype <- read.table("phenotype.txt",head=TRUE,sep="\t")
map <- read.table("mvp.plink.geno.map" , head = TRUE)

for(i in 2:ncol(phenotype)){
  imMVP <- MVP(
    phe=phenotype[, c(1, i)],
    geno=genotype,
    map=map,
    K=Kinship,
    #CV.GLM=Covariates,
    #CV.MLM=Covariates,
    #CV.FarmCPU=Covariates,
    nPC.GLM=3,
    nPC.MLM=3,
    nPC.FarmCPU=3,
    priority="speed",
    #ncpus=10,
    vc.method="BRENT",
    maxLoop=10,
    method.bin="static",
    #permutation.threshold=TRUE,
    #permutation.rep=100,
    threshold=0.05,
    method=c("GLM", "MLM", "FarmCPU"),
    p.threshold=0.00001
    
  )
  gc()
}

#########Phenotype distribution
MVP.Hist(phe=phenotype, file.type="jpg", breakNum=18, dpi=300)

#######GWAS OUT
data("Plant.Height")

##plot PCA
pca <- attach.big.matrix("mvp.vcf.pc.desc")[, 1:3]
#pca <- prcomp(t(as.matrix(genotype)))$x[, 1:3]
MVP.PCAplot(PCA=pca, Ncluster=3, class=NULL, col=c("red", "green", "yellow"), file.type="jpg")


```














